File: android/src/main/java/expo/modules/colorthief/ExpoColorThiefModule.kt
package expo.modules.colorthief

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.net.URL

class ExpoColorThiefModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoColorThief")

        AsyncFunction("getColor") { imageUri: String, quality: Int, ignoreWhite: Boolean, promise: Promise ->
            GlobalScope.launch(Dispatchers.IO) {
                try {
                    val bitmap = loadBitmap(imageUri)
                    if (bitmap == null) {
                        withContext(Dispatchers.Main) { promise.resolve(null) }
                        return@launch
                    }

                    val color = ColorThief.getColor(bitmap, quality, ignoreWhite)
                    withContext(Dispatchers.Main) {
                        if (color != null) {
                            promise.resolve(colorToMap(color))
                        } else {
                            promise.resolve(null)
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) { promise.resolve(null) }
                }
            }
        }

        AsyncFunction("getPalette") { imageUri: String, colorCount: Int, quality: Int, ignoreWhite: Boolean, promise: Promise ->
            GlobalScope.launch(Dispatchers.IO) {
                try {
                    val bitmap = loadBitmap(imageUri)
                    if (bitmap == null) {
                        withContext(Dispatchers.Main) { promise.resolve(null) }
                        return@launch
                    }

                    val palette = ColorThief.getPalette(bitmap, colorCount, quality, ignoreWhite)
                    withContext(Dispatchers.Main) {
                        if (palette != null) {
                            val colors = palette.map { colorToMap(it) }
                            promise.resolve(colors)
                        } else {
                            promise.resolve(null)
                        }
                    }
                } catch (e: Exception) {
                    withContext(Dispatchers.Main) { promise.resolve(null) }
                }
            }
        }
    }

    private fun colorToMap(color: IntArray): Map<String, Any> {
        val hex = String.format("#%02x%02x%02x", color[0], color[1], color[2])
        return mapOf(
            "rgb" to mapOf(
                "r" to color[0],
                "g" to color[1],
                "b" to color[2]
            ),
            "hex" to hex
        )
    }

    private fun loadBitmap(uri: String): Bitmap? {
        return try {
            when {
                uri.startsWith("http://") || uri.startsWith("https://") -> {
                    val url = URL(uri)
                    BitmapFactory.decodeStream(url.openStream())
                }
                uri.startsWith("file://") -> {
                    val path = uri.removePrefix("file://")
                    BitmapFactory.decodeFile(path)
                }
                uri.startsWith("content://") -> {
                    val context = appContext.reactContext ?: return null
                    val inputStream = context.contentResolver.openInputStream(Uri.parse(uri))
                    BitmapFactory.decodeStream(inputStream)
                }
                uri.startsWith("data:image") -> {
                    val base64 = uri.substringAfter(",")
                    val bytes = android.util.Base64.decode(base64, android.util.Base64.DEFAULT)
                    BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                }
                else -> {
                    // Try as file path
                    if (File(uri).exists()) {
                        BitmapFactory.decodeFile(uri)
                    } else {
                        null
                    }
                }
            }
        } catch (e: Exception) {
            null
        }
    }
}

File: ios/ExpoColorThiefModule.swift
import ExpoModulesCore
import UIKit

public class ExpoColorThiefModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoColorThief")

        AsyncFunction("getColor") { (imageUri: String, quality: Int, ignoreWhite: Bool, promise: Promise) in
            self.loadImage(from: imageUri) { image in
                guard let image = image else {
                    promise.resolve(nil)
                    return
                }

                guard let color = ColorThief.getColor(from: image, quality: quality, ignoreWhite: ignoreWhite) else {
                    promise.resolve(nil)
                    return
                }

                promise.resolve(self.colorToDict(color))
            }
        }

        AsyncFunction("getPalette") { (imageUri: String, colorCount: Int, quality: Int, ignoreWhite: Bool, promise: Promise) in
            self.loadImage(from: imageUri) { image in
                guard let image = image else {
                    promise.resolve(nil)
                    return
                }

                guard let palette = ColorThief.getPalette(from: image, colorCount: colorCount, quality: quality, ignoreWhite: ignoreWhite) else {
                    promise.resolve(nil)
                    return
                }

                let colors = palette.map { self.colorToDict($0) }
                promise.resolve(colors)
            }
        }
    }

    private func colorToDict(_ color: MMCQ.Color) -> [String: Any] {
        let hex = String(format: "#%02x%02x%02x", color.r, color.g, color.b)
        return [
            "rgb": [
                "r": Int(color.r),
                "g": Int(color.g),
                "b": Int(color.b)
            ],
            "hex": hex
        ]
    }

    private func loadImage(from uri: String, completion: @escaping (UIImage?) -> Void) {
        // Handle different URI schemes
        if uri.hasPrefix("file://") {
            let path = String(uri.dropFirst(7))
            let image = UIImage(contentsOfFile: path)
            completion(image)
            return
        }

        if uri.hasPrefix("http://") || uri.hasPrefix("https://") {
            guard let url = URL(string: uri) else {
                completion(nil)
                return
            }

            URLSession.shared.dataTask(with: url) { data, _, error in
                DispatchQueue.main.async {
                    if let data = data, error == nil {
                        completion(UIImage(data: data))
                    } else {
                        completion(nil)
                    }
                }
            }.resume()
            return
        }

        // Handle asset URIs (e.g., from require() or asset library)
        if uri.hasPrefix("asset://") || uri.hasPrefix("assets-library://") {
            // For asset library URIs, we need to use Photos framework
            // This is a simplified implementation
            completion(nil)
            return
        }

        // Handle base64 data URIs
        if uri.hasPrefix("data:image") {
            if let commaIndex = uri.firstIndex(of: ",") {
                let base64String = String(uri[uri.index(after: commaIndex)...])
                if let data = Data(base64Encoded: base64String) {
                    completion(UIImage(data: data))
                    return
                }
            }
            completion(nil)
            return
        }

        // Try as local file path
        if FileManager.default.fileExists(atPath: uri) {
            let image = UIImage(contentsOfFile: uri)
            completion(image)
            return
        }

        // Try as bundle resource
        if let image = UIImage(named: uri) {
            completion(image)
            return
        }

        completion(nil)
    }
}